File: apps/web/src/types/previewModel.ts
export type PreviewModel = {
  version: 1;
  name: string;
  format: "YDR" | "YFT";
  meshes: PreviewMesh[];
  textures: PreviewTexture[];
  bounds: {
    min: [number, number, number];
    max: [number, number, number];
  };
};

export type PreviewMesh = {
  name: string;
  positions: number[];
  normals: number[];
  uvs: number[];
  indices: number[];
  texture: string | null;
  shader: string;
  renderBucket: number;
};

export type PreviewTexture = {
  name: string;
  dds: string;
};

File: Engine/Contracts/PreviewModel.cs
namespace FiveMesh.Engine.Contracts;

internal sealed record PreviewModel(
    int Version,
    string Name,
    string Format,
    IReadOnlyList<PreviewMesh> Meshes,
    IReadOnlyList<PreviewTexture> Textures,
    PreviewBounds Bounds
);

internal sealed record PreviewMesh(
    string Name,
    float[] Positions,
    float[] Normals,
    float[] Uvs,
    int[] Indices,
    string? Texture,
    string Shader,
    byte RenderBucket
);

internal sealed record PreviewTexture(string Name, string Dds);

internal sealed record PreviewBounds(float[] Min, float[] Max);
